File: backend/src/controllers/taskController.ts
import { Request, Response } from 'express';
import { TaskService, TaskFilters, CreateTaskData, UpdateTaskData } from '../services/taskService';
import { validationResult } from 'express-validator';

const taskService = new TaskService();

export class TaskController {
  async getTasks(req: Request, res: Response) {
    try {
      const userId = (req as any).user.userId;
      const filters: TaskFilters = {
        completed: req.query.completed === 'true' ? true : req.query.completed === 'false' ? false : undefined,
        search: req.query.search as string,
        category: req.query.category as string,
        priority: req.query.priority as 'LOW' | 'MEDIUM' | 'HIGH',
        page: req.query.page ? parseInt(req.query.page as string) : undefined,
        limit: req.query.limit ? parseInt(req.query.limit as string) : undefined,
        sortBy: req.query.sortBy as 'deadline' | 'createdAt' | 'priority',
        sortOrder: req.query.sortOrder as 'asc' | 'desc',
      };

      const result = await taskService.getTasks(userId, filters);

      res.json({
        success: true,
        data: result,
      });
    } catch (error: any) {
      res.status(500).json({
        success: false,
        message: error.message || 'Failed to fetch tasks',
      });
    }
  }

  async createTask(req: Request, res: Response) {
    try {
      const errors = validationResult(req);
      if (!errors.isEmpty()) {
        return res.status(400).json({
          success: false,
          message: 'Validation failed',
          errors: errors.array(),
        });
      }

      const userId = (req as any).user.userId;
      const taskData: CreateTaskData = {
        title: req.body.title,
        description: req.body.description,
        category: req.body.category,
        priority: req.body.priority,
        deadline: req.body.deadline ? new Date(req.body.deadline) : undefined,
      };

      const task = await taskService.createTask(userId, taskData);

      res.status(201).json({
        success: true,
        message: 'Task created successfully',
        data: task,
      });
    } catch (error: any) {
      res.status(400).json({
        success: false,
        message: error.message || 'Failed to create task',
      });
    }
  }

  async updateTask(req: Request, res: Response) {
    try {
      const errors = validationResult(req);
      if (!errors.isEmpty()) {
        return res.status(400).json({
          success: false,
          message: 'Validation failed',
          errors: errors.array(),
        });
      }

      const userId = (req as any).user.userId;
      const taskId = parseInt(req.params.id);
      const taskData: UpdateTaskData = {
        title: req.body.title,
        description: req.body.description,
        category: req.body.category,
        priority: req.body.priority,
        completed: req.body.completed,
        deadline: req.body.deadline ? new Date(req.body.deadline) : undefined,
      };

      const task = await taskService.updateTask(taskId, userId, taskData);

      res.json({
        success: true,
        message: 'Task updated successfully',
        data: task,
      });
    } catch (error: any) {
      res.status(400).json({
        success: false,
        message: error.message || 'Failed to update task',
      });
    }
  }

  async deleteTask(req: Request, res: Response) {
    try {
      const userId = (req as any).user.userId;
      const taskId = parseInt(req.params.id);

      await taskService.deleteTask(taskId, userId);

      res.json({
        success: true,
        message: 'Task deleted successfully',
      });
    } catch (error: any) {
      res.status(400).json({
        success: false,
        message: error.message || 'Failed to delete task',
      });
    }
  }

  async toggleTask(req: Request, res: Response) {
    try {
      const userId = (req as any).user.userId;
      const { id } = req.params;
      const taskId = parseInt(id);

      const task = await taskService.toggleTask(taskId, userId);

      res.json({
        success: true,
        message: 'Task status toggled successfully',
        data: task,
      });
    } catch (error: any) {
      if (error.message === 'Task not found') {
        return res.status(404).json({
          success: false,
          message: error.message,
        });
      }

      res.status(500).json({
        success: false,
        message: error.message || 'Failed to toggle task',
      });
    }
  }

  async getTaskStats(req: Request, res: Response) {
    try {
      const userId = (req as any).user.userId;
      const stats = await taskService.getTaskStats(userId);

      res.json({
        success: true,
        data: stats,
      });
    } catch (error: any) {
      res.status(500).json({
        success: false,
        message: error.message || 'Failed to fetch task statistics',
      });
    }
  }
}

File: backend/src/main/java/com/taskmanager/controller/TaskController.java
package com.taskmanager.controller;

import com.taskmanager.dto.ApiResponse;
import com.taskmanager.dto.PaginatedTasksResponse;
import com.taskmanager.dto.TaskRequest;
import com.taskmanager.dto.TaskStatsResponse;
import com.taskmanager.model.Task;
import com.taskmanager.security.UserPrincipal;
import com.taskmanager.service.TaskService;
import jakarta.validation.Valid;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.*;

@RestController
@RequestMapping("/api/tasks")
public class TaskController {

    private final TaskService taskService;

    public TaskController(TaskService taskService) {
        this.taskService = taskService;
    }

    @GetMapping
    public ResponseEntity<ApiResponse<PaginatedTasksResponse>> getTasks(
            @AuthenticationPrincipal UserPrincipal principal,
            @RequestParam(required = false) Boolean completed,
            @RequestParam(required = false) String search,
            @RequestParam(required = false) String category,
            @RequestParam(required = false) String priority,
            @RequestParam(required = false) Integer page,
            @RequestParam(required = false) Integer limit,
            @RequestParam(required = false) String sortBy,
            @RequestParam(required = false) String sortOrder) {

        PaginatedTasksResponse response = taskService.getTasks(
                principal.getUserId(), completed, search, category, priority, page, limit, sortBy, sortOrder
        );
        return ResponseEntity.ok(new ApiResponse<>(true, response));
    }

    @PostMapping
    public ResponseEntity<ApiResponse<Task>> createTask(
            @AuthenticationPrincipal UserPrincipal principal,
            @Valid @RequestBody TaskRequest request) {

        Task task = taskService.createTask(principal.getUserId(), request);
        ApiResponse<Task> apiResponse = new ApiResponse<>(true, "Task created successfully", task);
        return ResponseEntity.status(HttpStatus.CREATED).body(apiResponse);
    }

    @PatchMapping("/{id}")
    public ResponseEntity<ApiResponse<Task>> updateTask(
            @PathVariable Integer id,
            @AuthenticationPrincipal UserPrincipal principal,
            @Valid @RequestBody TaskRequest request) {

        Task task = taskService.updateTask(id, principal.getUserId(), request);
        ApiResponse<Task> apiResponse = new ApiResponse<>(true, "Task updated successfully", task);
        return ResponseEntity.ok(apiResponse);
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<ApiResponse<Void>> deleteTask(
            @PathVariable Integer id,
            @AuthenticationPrincipal UserPrincipal principal) {

        taskService.deleteTask(id, principal.getUserId());
        ApiResponse<Void> apiResponse = new ApiResponse<>(true, "Task deleted successfully");
        return ResponseEntity.ok(apiResponse);
    }

    @PatchMapping("/{id}/toggle")
    public ResponseEntity<ApiResponse<Task>> toggleTask(
            @PathVariable Integer id,
            @AuthenticationPrincipal UserPrincipal principal) {

        Task task = taskService.toggleTask(id, principal.getUserId());
        ApiResponse<Task> apiResponse = new ApiResponse<>(true, "Task status toggled successfully", task);
        return ResponseEntity.ok(apiResponse);
    }

    @GetMapping("/stats")
    public ResponseEntity<ApiResponse<TaskStatsResponse>> getTaskStats(
            @AuthenticationPrincipal UserPrincipal principal) {

        TaskStatsResponse stats = taskService.getTaskStats(principal.getUserId());
        return ResponseEntity.ok(new ApiResponse<>(true, stats));
    }
}
